File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// O(n) time complexity and O(1) space complexity
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> ans;
        auto cur = root;
        while(cur!=nullptr){
            if(cur->left==nullptr){
                ans.push_back(cur->val);
                cur = cur->right;
            }
            else{
                auto prev = cur->left;
                while(prev->right&&prev->right!=cur){
                    prev = prev->right;
                }
                if(prev->right==nullptr){
                    prev->right = cur;
                    cur = cur->left;
                }
                else{
                    ans.push_back(cur->val);
                    prev->right = nullptr;
                    cur = cur->right;
                }
            }
        }
        return ans;
    }
};

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        temp = root
        ans = []
        while temp:
            if not temp.left:
                ans.append(temp.val)
                temp = temp.right
            else:
                rightmostLeaf = temp.left
                while rightmostLeaf.right and rightmostLeaf.right!=temp:
                    rightmostLeaf = rightmostLeaf.right
                
                if rightmostLeaf.right==None:
                    rightmostLeaf.right = temp
                    temp = temp.left
                else:
                    ans.append(temp.val)
                    rightmostLeaf.right = None
                    temp = temp.right
        return ans